File: modules/angular2/src/core/testability/testability.ts
import {Injectable} from 'angular2/src/core/di';
import {Map, MapWrapper, ListWrapper} from 'angular2/src/facade/collection';
import {CONST, CONST_EXPR, scheduleMicroTask} from 'angular2/src/facade/lang';
import {BaseException} from 'angular2/src/facade/exceptions';
import {NgZone} from '../zone/ng_zone';
import {ObservableWrapper} from 'angular2/src/facade/async';


/**
 * The Testability service provides testing hooks that can be accessed from
 * the browser and by services such as Protractor. Each bootstrapped Angular
 * application on the page will have an instance of Testability.
 */
@Injectable()
export class Testability {
  /** @internal */
  _pendingCount: number = 0;
  /** @internal */
  _isZoneStable: boolean = true;
  /**
   * Whether any work was done since the last 'whenStable' callback. This is
   * useful to detect if this could have potentially destabilized another
   * component while it is stabilizing.
   * @internal
   */
  _didWork: boolean = false;
  /** @internal */
  _callbacks: Function[] = [];
  constructor(private _ngZone: NgZone) { this._watchAngularEvents(); }

  /** @internal */
  _watchAngularEvents(): void {
    ObservableWrapper.subscribe(this._ngZone.onUnstable, (_) => {
      this._didWork = true;
      this._isZoneStable = false;
    });

    this._ngZone.runOutsideAngular(() => {
      ObservableWrapper.subscribe(this._ngZone.onStable, (_) => {
        NgZone.assertNotInAngularZone();
        scheduleMicroTask(() => {
          this._isZoneStable = true;
          this._runCallbacksIfReady();
        });
      });
    });
  }

  increasePendingRequestCount(): number {
    this._pendingCount += 1;
    this._didWork = true;
    return this._pendingCount;
  }

  decreasePendingRequestCount(): number {
    this._pendingCount -= 1;
    if (this._pendingCount < 0) {
      throw new BaseException('pending async requests below zero');
    }
    this._runCallbacksIfReady();
    return this._pendingCount;
  }

  isStable(): boolean {
    return this._isZoneStable && this._pendingCount == 0 && !this._ngZone.hasPendingMacrotasks;
  }

  /** @internal */
  _runCallbacksIfReady(): void {
    if (this.isStable()) {
      // Schedules the call backs in a new frame so that it is always async.
      scheduleMicroTask(() => {
        while (this._callbacks.length !== 0) {
          (this._callbacks.pop())(this._didWork);
        }
        this._didWork = false;
      });
    } else {
      // Not Ready
      this._didWork = true;
    }
  }

  whenStable(callback: Function): void {
    this._callbacks.push(callback);
    this._runCallbacksIfReady();
  }

  getPendingRequestCount(): number { return this._pendingCount; }

  findBindings(using: any, provider: string, exactMatch: boolean): any[] {
    // TODO(juliemr): implement.
    return [];
  }

  findProviders(using: any, provider: string, exactMatch: boolean): any[] {
    // TODO(juliemr): implement.
    return [];
  }
}

/**
 * A global registry of {@link Testability} instances for specific elements.
 */
@Injectable()
export class TestabilityRegistry {
  /** @internal */
  _applications = new Map<any, Testability>();

  constructor() { _testabilityGetter.addToWindow(this); }

  registerApplication(token: any, testability: Testability) {
    this._applications.set(token, testability);
  }

  getTestability(elem: any): Testability { return this._applications.get(elem); }

  getAllTestabilities(): Testability[] { return MapWrapper.values(this._applications); }

  getAllRootElements(): any[] { return MapWrapper.keys(this._applications); }

  findTestabilityInTree(elem: Node, findInAncestors: boolean = true): Testability {
    return _testabilityGetter.findTestabilityInTree(this, elem, findInAncestors);
  }
}

/**
 * Adapter interface for retrieving the `Testability` service associated for a
 * particular context.
 */
export interface GetTestability {
  addToWindow(registry: TestabilityRegistry): void;
  findTestabilityInTree(registry: TestabilityRegistry, elem: any,
                        findInAncestors: boolean): Testability;
}

@CONST()
class _NoopGetTestability implements GetTestability {
  addToWindow(registry: TestabilityRegistry): void {}
  findTestabilityInTree(registry: TestabilityRegistry, elem: any,
                        findInAncestors: boolean): Testability {
    return null;
  }
}

/**
 * Set the {@link GetTestability} implementation used by the Angular testing framework.
 */
export function setTestabilityGetter(getter: GetTestability): void {
  _testabilityGetter = getter;
}

var _testabilityGetter: GetTestability = CONST_EXPR(new _NoopGetTestability());

File: lib/src/core/testability/testability.dart
library angular2.src.core.testability.testability;

import "package:angular2/src/core/di.dart" show Injectable;
import "package:angular2/src/facade/collection.dart"
    show Map, MapWrapper, ListWrapper;
import "package:angular2/src/facade/exceptions.dart"
    show BaseException, WrappedException;
import "../zone/ng_zone.dart" show NgZone;
import "package:angular2/src/facade/async.dart"
    show PromiseWrapper, ObservableWrapper;

/**
 * The Testability service provides testing hooks that can be accessed from
 * the browser and by services such as Protractor. Each bootstrapped Angular
 * application on the page will have an instance of Testability.
 */
@Injectable()
class Testability {
  /** @internal */
  num _pendingCount = 0;
  /** @internal */
  List<Function> _callbacks = [];
  /** @internal */
  bool _isAngularEventPending = false;
  Testability(NgZone _ngZone) {
    this._watchAngularEvents(_ngZone);
  }
  /** @internal */
  void _watchAngularEvents(NgZone _ngZone) {
    ObservableWrapper.subscribe(_ngZone.onTurnStart, (_) {
      this._isAngularEventPending = true;
    });
    _ngZone.runOutsideAngular(() {
      ObservableWrapper.subscribe(_ngZone.onEventDone, (_) {
        if (!_ngZone.hasPendingTimers) {
          this._isAngularEventPending = false;
          this._runCallbacksIfReady();
        }
      });
    });
  }

  num increasePendingRequestCount() {
    this._pendingCount += 1;
    return this._pendingCount;
  }

  num decreasePendingRequestCount() {
    this._pendingCount -= 1;
    if (this._pendingCount < 0) {
      throw new BaseException("pending async requests below zero");
    }
    this._runCallbacksIfReady();
    return this._pendingCount;
  }

  bool isStable() {
    return this._pendingCount == 0 && !this._isAngularEventPending;
  }

  /** @internal */
  void _runCallbacksIfReady() {
    if (!this.isStable()) {
      return;
    }
    // Schedules the call backs in a new frame so that it is always async.
    PromiseWrapper.resolve(null).then((_) {
      while (!identical(this._callbacks.length, 0)) {
        (this._callbacks.removeLast())();
      }
    });
  }

  void whenStable(Function callback) {
    this._callbacks.add(callback);
    this._runCallbacksIfReady();
  }

  num getPendingRequestCount() {
    return this._pendingCount;
  }
  // This only accounts for ngZone, and not pending counts. Use `whenStable` to

  // check for stability.
  bool isAngularEventPending() {
    return this._isAngularEventPending;
  }

  List<dynamic> findBindings(dynamic using, String provider, bool exactMatch) {
    // TODO(juliemr): implement.
    return [];
  }

  List<dynamic> findProviders(dynamic using, String provider, bool exactMatch) {
    // TODO(juliemr): implement.
    return [];
  }
}

/**
 * A global registry of [Testability] instances for specific elements.
 */
@Injectable()
class TestabilityRegistry {
  /** @internal */
  var _applications = new Map<dynamic, Testability>();
  TestabilityRegistry() {
    _testabilityGetter.addToWindow(this);
  }
  registerApplication(dynamic token, Testability testability) {
    this._applications[token] = testability;
  }

  Testability getTestability(dynamic elem) {
    return this._applications[elem];
  }

  List<Testability> getAllTestabilities() {
    return MapWrapper.values(this._applications);
  }

  Testability findTestabilityInTree(dynamic elem,
      [bool findInAncestors = true]) {
    return _testabilityGetter.findTestabilityInTree(
        this, elem, findInAncestors);
  }
}

/**
 * Adapter interface for retrieving the `Testability` service associated for a
 * particular context.
 */
abstract class GetTestability {
  void addToWindow(TestabilityRegistry registry);
  Testability findTestabilityInTree(
      TestabilityRegistry registry, dynamic elem, bool findInAncestors);
}

class _NoopGetTestability implements GetTestability {
  void addToWindow(TestabilityRegistry registry) {}
  Testability findTestabilityInTree(
      TestabilityRegistry registry, dynamic elem, bool findInAncestors) {
    return null;
  }

  const _NoopGetTestability();
}

/**
 * Set the [GetTestability] implementation used by the Angular testing framework.
 */
void setTestabilityGetter(GetTestability getter) {
  _testabilityGetter = getter;
}

GetTestability _testabilityGetter = const _NoopGetTestability();
